### PR TITLE
Add kodo deploy command

### LINKED ISSUE #11: Add a command to deploy an image
As a user I should be able to use the below command to deploy an image. 

```
kodo deploy --image=openshift/hello-openshift:latest --replicas=10 --port=80 --token=xyz --server=https://url  --n=mynamespace
```

Behind the scenes, this should create a new [`Deployment` ](https://docs.openshift.com/container-platform/4.1/applications/deployments/what-deployments-are.html#deployments-kube-deployments_what-deployments-are).

Something similar to this should be created

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-openshift
spec:
  replicas: 10
  selector:
    matchLabels:
      app: hello-openshift
  template:
    metadata:
      labels:
        app: hello-openshift
    spec:
      containers:
      - name: hello-openshift
        image: openshift/hello-openshift:latest
        ports:
        - containerPort: 80
```

You don't necessarily need to write a yaml, you could create a instance of the `Deployment` struct, and then call the API to create this Deployment.

Usage hint
```
go get k8s.io/api@v0.18.3
```
```
import "k8s.io/api/core/v1"
import appsv1 "k8s.io/api/apps/v1" //  alias this as appsv1
...
...
...

deploymentObj := &appsv1.Deployment{
		ObjectMeta: metav1.ObjectMeta{
			Name: "some-unique-name",
		},
		Spec: appsv1.DeploymentSpec{
			Replicas: &replicas, 
			Selector: &metav1.LabelSelector{
				MatchLabels: map[string]string{
					"app": "some-unique-name",
				},
			},
			Template: apiv1.PodTemplateSpec{
				ObjectMeta: metav1.ObjectMeta{
					Labels: map[string]string{
						"some-unique-name", 
					},
				},
				Spec: apiv1.PodSpec{
					Containers: []apiv1.Container{
						{
							Name:  "doesnt-matteer",
							Image: "image-to-be-deployed", // should come from flag
							Ports: []apiv1.ContainerPort{
								{
									Name:          "http",
									Protocol:      apiv1.ProtocolTCP,
									ContainerPort: 80, // should come from flag
								},
							},
						},
					},
				},
			},
		},
	}
	clientset.AppsV1().Deployments(NAMESPACE).Create(deploymentObj)

```

### PR DESCRIPTION
**Pull request by jaideepr97**
Fixes #11 

Summary
----------------------
Allow a kodo user to deploy an image to an OpenShift cluster, with namespace awareness

```
./bin/kodo deploy --image=openshift/hello-openshift:latest --replicas=10 --port=80 --token=TOKEN --server=CLUSTER --namespace=jaideep-test-june-18

```

Technical changes
-------------------------

List of changes in this PR

- created structs for environment variables and deployment related variables instead of using global variables - and passed references to them 

- added new flags for --image, --port, --replicas

- added deploy functionality in deploy.go

- Included error handling wherever required and addressed comments raised in previous PR

- Included namespace filtering while retrieving list of pods and creating deployment

- Other minor changes pointed out in the previous PR

------------------------------------------- 

deployment-console-1
deployment-console-2

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -23,19 +24,21 @@ func init() {
 	rcommand.AddCommand(listCommand)
 	rcommand.PersistentFlags().StringVarP(&cmd.Host, "server", "s", "myurl", "this is the cluster url")
 	rcommand.PersistentFlags().StringVarP(&cmd.Bearertoken, "token", "t", "usertoken", "this is the user token")
-	rcommand.PersistentFlags().StringVarP(&cmd.Namespace, "namespace", "n", "namespace", "this is the namespace")
+	rcommand.PersistentFlags().StringVarP(&cmd.Namespace, "namespace", "n", "", "this is the namespace")
```
Comment (sbose78) [inline main.go]: Could you add a comment or a note in the description specifying the implications of this change?

File: main.go [hunk 2]
```diff
@@ -23,19 +24,21 @@ func init() {
...
 	rcommand.MarkFlagRequired("server")
 }
 
 func main() {
+	argsWithProg := os.Args
+	fmt.Println(argsWithProg)
```
Comment (sbose78) [inline main.go]: If this is for better logging, please add some text before/after too.

Also, call this out in the PR description with a sample output.

File: main.go [hunk 3]
```diff
@@ -23,19 +24,21 @@ func init() {
...
 	rcommand.Execute()
 }
 
 var listCommand = &cobra.Command{
 	Use: "list",
 	Run: func(cm *cobra.Command, args []string) {
-		fmt.Println("List All Kubernetes Applications")
-		fmt.Printf("Fetching all applications from %s in namespace %s", cmd.Host, cmd.Namespace)
+		fmt.Println("\nList All Kubernetes Applications")
```
Comment (sbose78) [inline main.go]: Why do we need the newline in Println(..)

File: pkg/kodo/cmd/list.go [hunk 4]
```diff
@@ -2,15 +2,20 @@ package cmd
...
 //List
 func List() error {
-	client := newOpenShiftClient()
-	pods, _ := client.CoreV1().Pods("").List(v1.ListOptions{})
-	fmt.Printf("\n The number of pods are %d \n", len(pods.Items))
-	return nil
+	client, clientError := newOpenShiftClient()
+
+	if clientError != nil {
+		log.Fatal(clientError)
```
Comment (sbose78) [inline pkg/kodo/cmd/list.go:16]: Should we continue trying to call the Pods() API if there were errors during client creation?

File: pkg/kodo/cmd/openshiftclient.go [hunk 5]
```diff
@@ -11,15 +11,15 @@ var (
...
+func newOpenShiftClient() (*kubernetes.Clientset, error) {
 	config := rest.Config{
 		Host:        Host,
 		BearerToken: Bearertoken,
 		TLSClientConfig: rest.TLSClientConfig{
 			Insecure: true,
 		},
 	}
-	myClientSet, _ := kubernetes.NewForConfig(&config)
+	myClientSet, clientsetError := kubernetes.NewForConfig(&config)
```
Comment (sbose78) [inline pkg/kodo/cmd/openshiftclient.go]: Maybe do a `return kubernetes.NewForConfig(...)`

File: test.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+placeholder file
```
Comment (sbose78) [inline test.txt]: This file isn't needed I guess :)

File: pkg/kodo/cmd/list.go [hunk 7]
```diff
@@ -2,15 +2,20 @@ package cmd
...
+	}
+	pods, podlisterror := client.CoreV1().Pods(Namespace).List(v1.ListOptions{})
+	fmt.Printf("\nThe number of pods are %d \n", len(pods.Items))
```
Comment (sbose78) [inline pkg/kodo/cmd/list.go]: Does it make sense to print this if podListerror isn't nil ? In other words, I wonder if printing zero pods in error scenarios might be misleading for users?

**Review by sbose78 — COMMENTED** [review]
Excellent improvements!

I have a few questions..

**Comment by sbose78** [comment]
Could you provide a full sample command which has worked ?

**Comment by jaideepr97** [comment]
./bin/kodo deploy --image=openshift/hello-openshift:latest --replicas=10 --port=80 --token=mpBBtlY8vQgCJWD1TOdCRU6LNfIQJDUPcEoXjkFrfyg --server=https://api.dev-svc-4.4-061813.devcluster.openshift.com:6443 --namespace=jaideep-test-june-18

**Comment by sbose78** [comment]
^ should be in the description.

**Comment by jaideepr97** [comment]
updated description

**Review by reginaelyse — APPROVED** [review]

**Review by sbose78 — APPROVED** [review]

**Pull request merged**